theories/structure/Subtype.eca

```diff
@@ -3,142 +3,105 @@ pragma +implicits.
+   [do_not_use_this_type_or_you_get_unsoundness_admit`, we make sure
+   that the user cannot do `type sT <- ...` when cloning. *) 
 type sT = [Wrap of do_not_use_this_type_or_you_get_unsoundness_admit].
 
-(* We define `val` in two steps so that `print val` does not show the ugly internal stuff *)
-op nosmt internal_val = choiceb (fun(f : do_not_use_this_type_or_you_get_unsoundness_admit -> T), injective f /\ (forall (x:T), ((P x) <=> exists y, x = f y))) witness.
+(* We define `val` in two steps so that `print val` does not show the
+   ugly internal stuff *)
+op nosmt internal_val = choiceb 
+   (fun (f : do_not_use_this_type_or_you_get_unsoundness_admit -> T), 
+     injective f /\ (forall (x:T), ((P x) <=> exists y, x = f y))) witness.
 
-(* Injection from `sT` to `T` *)
-op nosmt val (x : sT) : T = with x = Wrap x' => internal_val x'.
+(* -------------------------------------------------------------------- *)
 
-(* Inverse of a partial function. This and its related lemmas might be moved to somewhere else in the stdlib *)
-op pinverse (f : 'a -> 'b) (y : 'b) : 'a option = if exists x, y = f x then Some (choiceb (fun x, y = f x) witness) else None.
+(* Injection from [sT] to [T] *)
+op nosmt [opaque] val (x : sT) : T = with x = Wrap x' => internal_val x'.
 
 lemma val_range x: P x <=> exists y, x = val y.
-  have H: (fun(f : do_not_use_this_type_or_you_get_unsoundness_admit -> T), injective f /\ (forall (x:T), ((P x) <=> exists y, x = f y))) internal_val.
-    apply choicebP. by apply injection.
-  have H1: P x => exists (y : sT), x = val y.
-    move => Px.
-    have exy: exists y, x = internal_val y.
-      smt().
-    elim exy => z xz.
-    exists (Wrap z).
-    smt().
-  have H2: (exists (y : sT), x = val y) => P x.
-    move => ex.
-    elim ex => y.
-    elim y.
-    smt().
-  progress.
+proof.
+have /= := choicebP _ witness subtype_injection; rewrite -/internal_val.
+move => [ival_inj ival_range].
+split => [Px|[[y] ->]]; 2: smt().
+have := ival_range x; rewrite Px /= => -[z ?].
+by exists (Wrap z); smt().
 qed.
 
 lemma val_inj: injective val.
-  have H: (fun(f : do_not_use_this_type_or_you_get_unsoundness_admit -> T), injective f /\ (forall (x:T), ((P x) <=> exists y, x = f y))) internal_val.
-    apply choicebP. by apply injection.
-  have H1: injective internal_val.
-    smt().
-  clear H.
-  rewrite /injective => x y.
-  elim x. elim y.
-  rewrite /val.
-  smt().
+proof.
+have /= := choicebP _ witness subtype_injection; rewrite -/internal_val.
+by move => [ival_inj _] [x] [y] /#.
 qed.
 
-lemma valP (x : sT): P (val x).
-  smt (val_range).
-qed.
+lemma valP (x : sT): P (val x) by smt (val_range).
 
-op nosmt insub : T -> sT option = pinverse val.
+(* -------------------------------------------------------------------- *)
 
-lemma pinverseN (f:'a->'b) x: (!exists y, x = f y) => pinverse f x = None.
-  smt().
-qed.
+op nosmt [opaque] insub : T -> sT option = pinv val.
```

Comment:
@chdoc 
Maybe this should not be `[opaque]`? After all, the definition of insub is not just some strange encoding 
with some arbitrary choice (like that of `val`), but a very canonical definition in terms of the visible op `val`. So exposing this to the user makes sense to me.

Reply:
It's still available to the user via `rewrite /insub`, but just not used automatically by tactics.

```
op [opaque] x : int = 3 + 4.

lemma foo : x = 3 + 4.
proof.
trivial.
(*
Current goal

Type variables: <none>

------------------------------------------------------------------------
x = 3 + 4
*)
rewrite /x.
(*
Current goal

Type variables: <none>

------------------------------------------------------------------------
3 + 4 = 3 + 4
*)
```

Reply:
Ah, ok. Then I guess we can leave it that way.